fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java

```diff
@@ -0,0 +1,848 @@
+            } else {
+                return;
+            }
+        }
+
+        try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
+            LakeTable table = getTableOrThrow(db, tableId);
+            commitVersionMap = new HashMap<>();
+            for (long partitionId : partitionIndexMap.rowKeySet()) {
+                Partition partition = table.getPartition(partitionId);
+                Preconditions.checkNotNull(partition, partitionId);
+                long commitVersion = partition.getNextVersion();
+                commitVersionMap.put(partitionId, commitVersion);
+                LOG.debug("commit version of partition {} is {}. jobId={}", partitionId, commitVersion, jobId);
+            }
+            this.jobState = JobState.FINISHED_REWRITING;
+
+            writeEditLog(this);
+
+            // NOTE: !!! below this point, this schema change job must success unless the database or table been dropped. !!!
+            updateNextVersion(table);
+        }
+
+        if (span != null) {
+            span.addEvent("finishedRewriting");
+        }
+        LOG.info("schema change job finished rewriting historical data: {}", jobId);
+    }
+
+    // Note: The only allowed situation to cancel the schema change job is the table has been dropped.
+    @Override
+    protected void runFinishedRewritingJob() throws AlterCancelException {
+        Preconditions.checkState(jobState == JobState.FINISHED_REWRITING);
+        // If the table or database has been dropped, `readyToPublishVersion()` will throw AlterCancelException and
+        // this schema change job will be cancelled.
+        if (!readyToPublishVersion()) {
+            return;
+        }
+
+        if (!publishVersion()) {
+            LOG.info("publish version failed, will retry later. jobId={}", jobId);
+            return;
+        }
+
+        this.jobState = JobState.FINISHED;
+        this.finishedTimeMs = System.currentTimeMillis();
+
+        writeEditLog(this);
+
+        // Replace the current index with shadow index.
+        List<MaterializedIndex> droppedIndexes;
+        try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
+            LakeTable table = (db != null) ? db.getTable(tableId) : null;
+            if (table == null) {
+                LOG.info("database or table been dropped while doing schema change job {}", jobId);
+                return;
+            }
+            // Below this point, all query and load jobs will use the new schema.
+            droppedIndexes = visualiseShadowIndex(table);
+        }
+
+        // Delete shards from StarOS
+        try {
+            StarOSAgent starOSAgent = GlobalStateMgr.getCurrentStarOSAgent();
+            for (MaterializedIndex droppedIndex : droppedIndexes) {
+                starOSAgent.deleteShards(droppedIndex.getTablets().stream().map(Tablet::getId).collect(Collectors.toSet()));
```

Comment:
Why not use ShardDeleter?

ShardDeleter ensures that the staros shard will not be deleted until the tablet metadata is deleted.

Reply:
updated